libsmtutil/SMTLib2Context.cpp

```diff
@@ -0,0 +1,316 @@
+		smtAssert(it != m_bitVectorSorts.end());
+		return std::make_shared<BitVectorSort>(it->first);
+	}
+	case Kind::Array:
+	{
+		auto it = ranges::find_if(m_arraySorts, [&](auto const& entry) { return entry.second == _sort; });
+		smtAssert(it != m_arraySorts.end());
+		return std::make_shared<ArraySort>(unresolve(it->first.first), unresolve(it->first.second));
+	}
+	case Kind::Tuple:
+	{
+		auto const& tupleType = dynamic_cast<TupleType const&>(type);
+		std::vector<std::string> memberNames;
+		std::vector<SortPointer> memberTypes;
+		for (auto&& [name, sortId] : tupleType.m_accessors)
+		{
+			memberNames.push_back(name);
+			memberTypes.push_back(unresolve(sortId));
+		}
+		return std::make_shared<TupleSort>(tupleType.m_name, std::move(memberNames), std::move(memberTypes));
+	}
+	default:
+		smtAssert(false, "Invalid SMT sort");
+	}
+}
+
+SortId SMTLib2Context::resolveBitVectorSort(BitVectorSort const& _sort)
+{
+	auto size = _sort.size;
+	auto it = m_bitVectorSorts.find(size);
+	if (it == m_bitVectorSorts.end())
+	{
+		auto newId = static_cast<uint32_t>(m_knownTypes.size());
+		m_knownTypes.emplace_back(std::make_unique<SMTLibSort>(Kind::BitVector, "(_ BitVec " + std::to_string(size) + ')', std::vector<SortId>{}, SortId{newId}));
+		auto&& [newIt, inserted] = m_bitVectorSorts.emplace(size, SortId{newId});
+		smtAssert(inserted);
+		it = newIt;
+	}
+	return it->second;
+}
+
+SortId SMTLib2Context::resolveArraySort(ArraySort const& _sort)
+{
+	smtAssert(_sort.domain && _sort.range);
+	auto domainSort = resolve(_sort.domain);
+	auto rangeSort = resolve(_sort.range);
+	auto pair = std::make_pair(domainSort, rangeSort);
+	auto it = m_arraySorts.find(pair);
+	if (it == m_arraySorts.end())
+	{
+		auto newId = static_cast<uint32_t>(m_knownTypes.size());
+		m_knownTypes.emplace_back(std::make_unique<SMTLibSort>(Kind::Array, "Array", std::vector<SortId>{domainSort, rangeSort}, SortId{newId}));
+		auto&& [newIt, inserted] = m_arraySorts.emplace(pair, SortId{newId});
+		smtAssert(inserted);
+		it = newIt;
+	}
+	return it->second;
+}
+
+SortId SMTLib2Context::resolveTupleSort(TupleSort const& _sort)
+{
+	auto const& tupleName = _sort.name;
+	auto it = m_tupleSorts.find(tupleName);
+	if (it == m_tupleSorts.end())
+	{
+		std::vector<std::pair<std::string, SortId>> accessors;
+		smtAssert(_sort.members.size() == _sort.components.size());
+		for (std::size_t i = 0u; i < _sort.members.size(); ++i)
+			accessors.emplace_back(_sort.members[i], resolve(_sort.components[i]));
+		auto newId = static_cast<uint32_t>(m_knownTypes.size());
+		m_knownTypes.emplace_back(std::make_unique<TupleType>(tupleName, std::move(accessors), SortId{newId}));
+		auto&& [newIt, inserted] = m_tupleSorts.emplace(tupleName, SortId{newId});
+		smtAssert(inserted);
+		it = newIt;
+		if (m_callback)
+			m_callback(_sort);
+	}
+	return it->second;
+}
+
+std::string SMTLib2Context::toString(solidity::smtutil::SortId _id)
+{
+	auto const& sort = m_knownTypes.at(_id.id);
```

Comment:
If we used a typedef instead of custom struct we would have here just `_id` and not `_id.id` :+)